compiler/rustc_interface/src/interface.rs

```diff
@@ -336,16 +336,53 @@ pub fn run_compiler<R: Send>(config: Config, f: impl FnOnce(&Compiler) -> R + Se
 
             let early_dcx = EarlyDiagCtxt::new(config.opts.error_format);
 
-            let codegen_backend = if let Some(make_codegen_backend) = config.make_codegen_backend {
-                make_codegen_backend(&config.opts)
-            } else {
-                util::get_codegen_backend(
-                    &early_dcx,
-                    &config.opts.maybe_sysroot,
-                    config.opts.unstable_opts.codegen_backend.as_deref(),
-                )
+            let sysroot = filesearch::materialize_sysroot(config.opts.maybe_sysroot.clone());
+
+            let (codegen_backend, target_override) = match config.make_codegen_backend {
+                None => {
+                    // Build a target without override, so that it can override the backend if needed
+                    let target =
+                        config::build_target_config(&early_dcx, &config.opts, None, &sysroot);
+
+                    let backend = util::get_codegen_backend(
+                        &early_dcx,
+                        &sysroot,
+                        config.opts.unstable_opts.codegen_backend.as_deref(),
+                        &target,
+                    );
+
+                    // target_override is documented to be called before init(), so this is okay
+                    let target_override = backend.target_override(&config.opts);
```

Comment:
This doesn't really do the same thing as before, or rather it assumes custom codegen backends require `config.make_codegen_backend`, whereas `config.opts.unstable_opts.codegen_backend` is also possible.

The new version calls `config::build_target_config` once, without the override, which means the backend *cannot* add new targets, only at most *slightly tweak* existing ones.

To be clear, none of this really matters anymore due to:
- https://github.com/rust-lang/rust/pull/122810

(but I ran into it while trying to update Rust-GPU to `nightly-2024-03-21`, just before #122810, just so I can get everything else out of the way before tackling the `target_override` removal itself)